src/pybind/mgr/drain_osd/module.py

```diff
@@ -0,0 +1,333 @@
+            self.reweight_osds(self.emptying_osds)
+
+            # 3) check for osds to be empty
+            empty_osds = self.empty_osds(self.emptying_osds)
+
+            # remove osds that are marked as empty
+            self.emptying_osds = self.emptying_osds.difference(empty_osds)
+
+            # Use mgr_tick_period (default: 2) here just to illustrate
+            # consuming native ceph options.  Any real background work
+            # would presumably have some more appropriate frequency.
+            sleep_interval = self.mgr_tick_period
+            self.log.debug('Sleeping for %d seconds', sleep_interval)
+            self.event.wait(sleep_interval)
+            self.event.clear()
+
+    def shutdown(self):
+        """
+        This method is called by the mgr when the module needs to shut
+        down (i.e., when the serve() function needs to exit).
+        """
+        self.log.info('Stopping')
+        self.run = False
+        self.event.set()
+
+    def osds_not_in_cluster(self, osd_ids: List[int]) -> Set[int]:
+        self.log.info(f"Checking if provided osds <{osd_ids}> exist in the cluster")
+        osd_map = self.get_osdmap()
+        cluster_osds = [x.get('osd') for x in osd_map.dump().get('osds', [])]
+        not_in_cluster = set()
+        for osd_id in osd_ids:
+            if int(osd_id) not in cluster_osds:
+                self.log.error(f"Could not find {osd_id} in cluster")
+                not_in_cluster.add(osd_id)
+        return not_in_cluster
+
+    def empty_osds(self, osd_ids: Set[int]) -> List[int]:
+        if not osd_ids:
+            return list()
+        osd_df_data = self.osd_df()
+        empty_osds = list()
+        for osd_id in osd_ids:
+            if self.is_empty(osd_id, osd_df=osd_df_data):
+                empty_osds.append(osd_id)
+        return empty_osds
+
+    def osd_df(self) -> dict:
+        # TODO: this should be cached I think
+        base_cmd = 'osd df'
+        ret, out, err = self.mon_command({
+            'prefix': base_cmd,
+            'format': 'json'
+        })
+        return json.loads(out)
+
+    def is_empty(self, osd_id: int, osd_df: Optional[dict] = None) -> bool:
+        pgs = self.get_pg_count(osd_id, osd_df=osd_df)
+        if pgs != 0:
+            self.log.info(f"osd: {osd_id} still has {pgs} PGs.")
+            return False
+        self.log.info(f"osd: {osd_id} has no PGs anymore")
+        return True
+
+    def reweight_osds(self, osd_ids: Set[int]) -> bool:
+        results = []
+        for osd_id in osd_ids:
+            results.append(self.reweight_osd(osd_id))
+        return all(results)
+
+    def get_pg_count(self, osd_id: int, osd_df: Optional[dict] = None) -> int:
+        if not osd_df:
+            osd_df = self.osd_df()
+        osd_nodes = osd_df.get('nodes', [])
+        for osd_node in osd_nodes:
+            if osd_node.get('id', None) == int(osd_id):
+                return osd_node.get('pgs')
+        return -1
```

Comment:
Would it make sense to raise an exception here?

Reply:
Since I check for OSD existence _before_ this codepath is reached, I wouldn't ever return -1..
Also I'm not sure in which circumstances `osd df` can be empty.. Let me think about that..